sui_core/src/gateway_types.rs

```diff
@@ -0,0 +1,947 @@
+    MoveObject(SuiMoveObject),
+    Package(SuiMovePackage),
+}
+
+#[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, Eq, PartialEq)]
+#[serde(rename = "MoveObject")]
+pub struct SuiMoveObject {
+    #[serde(rename = "type")]
+    pub type_: String,
+    pub contents: SuiMoveValue,
+}
+
+impl TryFrom<&SuiObject> for GasCoin {
+    type Error = SuiError;
+    fn try_from(object: &SuiObject) -> Result<Self, Self::Error> {
+        match &object.data {
+            SuiData::MoveObject(o) => {
+                if let SuiMoveValue::Coin(coin) = &o.contents {
+                    return Ok(GasCoin(coin.clone()));
+                }
+            }
+            SuiData::Package(_) => {}
+        }
+
+        return Err(SuiError::TypeError {
+            error: format!(
+                "Gas object type is not a gas coin: {:?}",
+                object.data.type_()
+            ),
+        });
+    }
+}
+
+impl SuiData {
+    pub fn try_as_package(&self) -> Option<&SuiMovePackage> {
+        match self {
+            SuiData::MoveObject(_) => None,
+            SuiData::Package(p) => Some(p),
+        }
+    }
+    pub fn type_(&self) -> Option<&str> {
+        match self {
+            SuiData::MoveObject(m) => Some(&m.type_),
+            SuiData::Package(_) => None,
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, JsonSchema)]
+pub struct PublishResponse {
+    /// Certificate of the transaction
+    pub certificate: SuiCertifiedTransaction,
+    /// The newly published package object reference.
+    pub package: SuiObjectRef,
+    /// List of Move objects created as part of running the module initializers in the package
+    pub created_objects: Vec<SuiObject>,
+    /// The updated gas payment object after deducting payment
+    pub updated_gas: SuiObject,
+}
+
+impl Display for PublishResponse {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut writer = String::new();
+        writeln!(writer, "----- Certificate ----")?;
+        write!(writer, "{}", self.certificate)?;
+        writeln!(writer, "----- Publish Results ----")?;
+        writeln!(
+            writer,
+            "The newly published package object ID: {:?}",
+            self.package.object_id
+        )?;
+        if !self.created_objects.is_empty() {
+            writeln!(
+                writer,
+                "List of objects created by running module initializers:\n"
+            )?;
+            for obj in &self.created_objects {
+                writeln!(writer, "{}\n", obj)?;
+            }
+        }
+        let gas_coin = GasCoin::try_from(&self.updated_gas).map_err(fmt::Error::custom)?;
+        writeln!(writer, "Updated Gas : {}", gas_coin)?;
+        write!(f, "{}", writer)
+    }
+}
+
+#[derive(Serialize, Deserialize, JsonSchema)]
+#[serde(tag = "status", content = "details", rename = "ObjectRead")]
+pub enum SuiObjectRead {
+    Exists(SuiObject),
+    NotExists(ObjectID),
+    Deleted(ObjectRef),
```

Comment:
nit: can we use `SuiObjectRef` consistently in the API level and avoid using tuples in the RPC layer in general?
```suggestion
    Deleted(SuiObjectRef),
```